[ticket/16587] Replace Sami by Doctum

williamdes opened the pull request:
Checklist:

- [x] Correct branch: master for new features; 3.3.x & 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html), [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-16587



<!-- 
williamdes@williamdes:/mnt/Dev/@code-lts$ git clone git@github.com:code-lts/phpbb.git
Clonage dans 'phpbb'...
remote: Enumerating objects: 151, done.
remote: Counting objects: 100% (151/151), done.
remote: Compressing objects: 100% (101/101), done.
error: object a7fc765d7baa51f07cc83f0efd5643a61595bf30: zeroPaddedFilemode: contains zero-padded file modes
fatal: erreur de fsck dans l'objet empaqueté
fatal: échec de index-pack


git clone --config transfer.fsckobjects=false \
    --config receive.fsckobjects=false \
    --config fetch.fsckobjects=false \
    git@github.com:code-lts/phpbb.git

-->

williamdes commented:
cc @JoshyPHP  

I need to rebase when I get to my laptop

JoshyPHP commented:
@williamdes I'm just a third party contributor here, @marc1706 usually handles merge requests.

rubencm commented:
You should do a rebase to fix the conflict

rxu commented:
It would  be good also to make use of it via `doctum.phar` instead of dependency as Sami/Doctum documentation states.

marc1706 commented:
I also tend to agree with using the phar instead. Feel free to ask if you need any assistance with moving over to that.

With regards to your composer.lock changes (which should become obsolete after switching to the phar) I do have one suggestion for the future: Make sure to first run `composer install` and then only add or update what you're looking for, e.g. `composer update code-lts/doctum` in this case.

williamdes commented:
Rebased, I will push one commit to use the phar ASAP (one small change required on my phar distribution system) :)

rxu commented:
@williamdes You also forgot to rename `travis/check-sami-parse-errors.sh` to `travis/check-doctum-parse-errors.sh` ;)

williamdes commented:
It should be alright now ;)

williamdes commented:
Dear maintainers, please commit something similar:
```diff
diff --git a/git-tools/hooks/commit-msg b/git-tools/hooks/commit-msg
index b73005b34d..fc3a4fdd48 100755
--- a/git-tools/hooks/commit-msg
+++ b/git-tools/hooks/commit-msg
@@ -337,7 +337,7 @@ done
 # If EOF is expected exit cleanly
 echo "$expecting" | grep -q "eof" || (
        # Unexpected EOF, error
-       complain "Unexpected EOF encountered" >&2;
+       complain "Unexpected EOF encountered in ${expecting}" >&2;
        quit $ERR_EOF;
 ) && (
        # Do post scan checks
```

I did my commits and was very perplex and had to do some die and retry locally to fix this error :) 

Errors: https://travis-ci.org/github/phpbb/phpbb/jobs/723157326#L623

Understood and fixed with a git CLI rebase.
Note: I am using GitKraken

rxu commented on travis/check-doctum-parse-errors.sh:
```diff
@@ -16,19 +16,6 @@ NOTESTS=$3
-	then
-		exit 1
-	fi
+	(cd phpBB/vendor/bin/ && curl -O https://doctum.long-term.support/releases/latest/doctum.phar && chmod +x doctum.phar)
```
I'd suggest to put `doctum.phar` to the `build` folder permanently so it could be used not in tests only but also to create docs manually via command line.
Also `build/build.xml` still contains Sami references: https://github.com/code-lts/phpbb/blob/625a16732e47ffa858d804a9bb61a659e162df85/build/build.xml#L131
https://github.com/code-lts/phpbb/blob/625a16732e47ffa858d804a9bb61a659e162df85/build/build.xml#L137

cba467829b0d6af5dec5ba2ca3718f8ede8d6ab4 is already known and would help probably :)

williamdes replied:
> Also build/build.xml still contains Sami references

:facepalm: By bad, vscode ignored pattern..

All fixed, thank you for the initial review

rxu commented on build/build.xml:
```diff
@@ -128,13 +128,13 @@
 	<target name="docs">
 		<exec dir="."
-			command="phpBB/vendor/bin/sami.php update build/sami-checkout.conf.php"
+			command="php doctum.phar update build/sami-checkout.conf.php"
```
I'm sorry but there's Sami leftover again -> `build/sami-checkout.conf.php` is now `build/doctum-checkout.conf.php`, same below with `build/sami-all.conf.php`.

rxu commented:
Also it seems that `doctum.phar` used in this PR is build with `symfony/comsole` 4.3 or later which is using `stream_isatty()` in `/Output/StreamOutput.php`. `stream_isatty()` is available since PHP 7.2 so is not compatible with current tests which use PHP 7.1 as a minimum.
https://travis-ci.org/github/phpbb/phpbb/jobs/723412665#L598

williamdes commented:
Sorry for the 2 mistakes..
Working on a fix upstream to generate a PHP 7.1 compatible phar

williamdes commented:
Build succeeded !
cc @marc1706 

https://travis-ci.org/github/phpbb/phpbb/builds/723500331

marc1706 commented:
I've noticed that there are quite a few changes in composer.lock, most of which seem like they are not directly caused by just removing sami. Therefore, I have created a change on top of your PR to only include the removal of sami. You can integrate this change from my branch which is your PR plus this one additional change as commit: https://github.com/marc1706/phpbb/tree/ticket/16587

williamdes commented:
> I've noticed that there are quite a few changes in composer.lock, most of which seem like they are not directly caused by just removing sami. Therefore, I have created a change on top of your PR to only include the removal of sami. You can integrate this change from my branch which is your PR plus this one additional change as commit: https://github.com/marc1706/phpbb/tree/ticket/16587

Thank you, my composer surely did some unwanted changes.
I rebased and squashed your changes into my commit as 52cb565eef626c45ad01d91a2e1cde5e2a210281

rxu commented:
So only `sami/sami` got removed from `composer.lock` but not its exclusive dependencies also, things like pimple/pimple etc are still there.
Is it intended?

rxu commented:
AFAIK simply `php ../composer.phar remove sami/sami` should do the job (remove `sami/sami` and its dependencies excepting global ones), but should be confirmed by @marc1706 I guess.

marc1706 commented:
You're right, someone that messed up before. @williamdes added the refactored commit that does composer remove to my branch above.

williamdes commented:
> You're right, someone that messed up before. @williamdes added the refactored commit that does composer remove to my branch above.

Let me know if I need to do something, else my PR is ready ;)

marc1706 commented:
Just cherry-pick my latest commit from my branch and that should already reflect these last changes needed:
https://github.com/phpbb/phpbb/compare/3.3.x...marc1706:ticket/16587

williamdes commented:
> Just cherry-pick my latest commit from my branch and that should already reflect these last changes needed:
> [3.3.x...marc1706:ticket/16587](https://github.com/phpbb/phpbb/compare/3.3.x...marc1706:ticket/16587)

Done

rxu commented on travis/check-doctum-parse-errors.sh:
```diff
@@ -0,0 +1,32 @@
+then
+	if [ ! -f doctum.phar ]; then
+		# Download the latest release
+		curl -O https://doctum.long-term.support/releases/5.1/doctum.phar
```
We have `doctum.phar` in the package now, shouldn\t we use it instead of downloading the latest?

williamdes replied:
We only download if the file does not exist ;)
This an update script if you remove the file

rxu replied:
Well, it always should exist now :) But well.

marc1706 commented on build/doctum-checkout.conf.php:
```diff
@@ -0,0 +1,29 @@
+	->notPath('#^vendor/#')
+	->notPath('data');
+
+return new Doctum\Doctum($iterator, [
```
Running with doctum-all.conf.php currently does not work as this config is not defined as the variable `$config` but passed directly to the constructor (see also previous version of this with sami). Once you've changed that, running with doctum-all.conf.php should also work fine.

williamdes replied:
Sure, I missed that
Now fixed in 3be5ab3a5c87544c4dc7786cfb5c4dec609d0d0b, I added a comment to document this behavior

marc1706 reviewed (changes requested):
Apart from that, rest looks fine.

3D-I commented:
Wow, great job here @williamdes 
It hasn't been an easy thing as far as I can see. :+1:

williamdes commented:
Let me know when it is deployed :)
